src/clj/puppetlabs/http/client/async.clj

```diff
@@ -326,39 +137,80 @@
-        ;; available then send the response body asynchronously
-        consumer (StreamingAsyncResponseConsumer.
-                  (FnDeliverable.
-                   (fn
-                     [http-response]
-                     (complete-response result opts callback http-response http-context))))
-        ;; If an error occurs early in the request, the consumer may not get a chance to return the
-        ;; response using the FnDeliverable. This wrapper around the future-callback guarantees that
-        ;; happens. It also takes care of notifying the consumer of the final result.
-        streaming-complete-callback (reify
-                                      FutureCallback
-                                      (completed [this http-response]
-                                        (.setFinalResult consumer nil)
-                                        (.completed future-callback http-response))
-                                      (failed [this ex]
-                                        (.setFinalResult consumer ex)
-                                        (.failed future-callback ex))
-                                      (cancelled [this]
-                                        (.setFinalResult consumer nil)
-                                        (.cancelled future-callback)))]
-    (.execute client
-              (HttpAsyncMethods/create request)
-              consumer
-              streaming-complete-callback)))
+   response :- common/Response]
+  (if callback
+    (try
+      (callback response)
+      (catch Exception e
+        {:opts opts :error e}))
+    response))
+
+(schema/defn java-content-type->clj :- common/ContentType
+  [java-content-type]
+  (if java-content-type
+    {:mime-type (.getMimeType java-content-type)
+     :charset   (.getCharset java-content-type)}))
+
+(schema/defn get-response-delivery-delegate :- ResponseDeliveryDelegate
+  [opts :- common/UserRequestOptions
+   result :- common/ResponsePromise]
+  (reify ResponseDeliveryDelegate
+    (deliverResponse
+      [_ _ orig-encoding body headers status content-type callback]
+      (let [response {:opts                  opts
+                      :orig-content-encoding orig-encoding
+                      :status                status
+                      :headers               (into {} headers)
+                      :content-type          (java-content-type->clj content-type)
+                      :body                  body}
+            response (callback-response opts callback response)]
+        (deliver result response)))
+    (deliverResponse
+      [_ _ e]
+      (deliver result {:opts opts :error e}))))
+
+(schema/defn clojure-method->java
+  [opts :- common/UserRequestOptions]
+  (case (:method opts)
+    :delete HttpMethod/DELETE
+    :get HttpMethod/GET
+    :head HttpMethod/HEAD
+    :options HttpMethod/OPTIONS
+    :patch HttpMethod/PATCH
+    :post HttpMethod/POST
+    :put HttpMethod/PUT
+    :trace HttpMethod/TRACE))
```

Comment:
Okay, that seems fine. Seems like it might be more possible to add a test after Alex's request PR is merged, but doesn't seem entirely necessary or worth the effort.